StenosisMeasurement3D/qSlicerStenosisMeasurement3DModuleWidget.cxx

```diff
@@ -355,56 +428,294 @@ void qSlicerStenosisMeasurement3DModuleWidget::onTubePointEndInteraction(vtkObje
+  {
+    // Disconnect the currently observed node.
+    this->inputFiducialNode->RemoveObserver(this->fiducialObservation);
+  }
+  if (this->inputShapeNode)
+  {
+    // Disconnect the currently observed node.
+    this->inputShapeNode->RemoveObserver(this->tubeObservation);
+  }
+  vtkMRMLStenosisMeasurement3DParameterNode* parameterNode = vtkMRMLStenosisMeasurement3DParameterNode::SafeDownCast(node);
+  d->ParameterNode = parameterNode;
+  d->EnableWorkspace();
+  this->updateWidgetFromMRML();
+}
+
+//----------------------------------------------------------------------------
+void qSlicerStenosisMeasurement3DModuleWidget::updateWidgetFromMRML()
+{
+  Q_D(qSlicerStenosisMeasurement3DModuleWidget);
+  if (!d->ParameterNode)
+  {
+    // Reset all UI widgets.
+    d->inputShapeSelector->setCurrentNode(nullptr);
+    d->inputFiducialSelector->setCurrentNode(nullptr);
+    d->inputSegmentSelector->setCurrentNode(nullptr);
+    d->inputSegmentSelector->setCurrentSegmentID(nullptr);
+    d->wallModelSelector->setCurrentNode(nullptr);
+    d->lumenModelSelector->setCurrentNode(nullptr);
+    d->outputTableSelector->setCurrentNode(nullptr);
+    d->wallResultLabel->clear();
+    d->lumenResultLabel->clear();
+    d->lesionResultLabel->clear();
+    d->stenosisResultLabel->clear();
+    d->lengthResultLabel->clear();
+    return;
+  }
+  
+  /*
+   * The parameter node will get in turn updated.
+   * Since we do not observe its changes because it's private,
+   * we will not end in infinite recursion.
+   * 
+   * The segmentation selector is special. The segmentID must be explicitly
+   * cleared if the segmentation is null.
+   */
+  d->inputShapeSelector->setCurrentNode(d->ParameterNode->GetInputShapeNode());
+  d->inputFiducialSelector->setCurrentNode(d->ParameterNode->GetInputFiducialNode());
+  bool wasBlocked = d->inputSegmentSelector->blockSignals(true);
+  d->inputSegmentSelector->setCurrentNode(d->ParameterNode->GetInputSegmentationNode());
+  d->inputSegmentSelector->setCurrentSegmentID(d->ParameterNode->GetInputSegmentationNode()
+                                    ? QString(d->ParameterNode->GetInputSegmentID())
+                                    : nullptr);
+  d->inputSegmentSelector->blockSignals(wasBlocked);
+  
+  d->wallModelSelector->setCurrentNode(d->ParameterNode->GetOutputWallModelNode());
+  d->lumenModelSelector->setCurrentNode(d->ParameterNode->GetOutputLumenModelNode());
+  d->outputTableSelector->setCurrentNode(d->ParameterNode->GetOutputTableNode());
+  
+  /*
+   * Despite the MRML table, recompute is unavoidable since it stores
+   * results from one or many parameter sets.
+  */
+  d->wallResultLabel->clear();
+  d->lumenResultLabel->clear();
+  d->lesionResultLabel->clear();
+  d->stenosisResultLabel->clear();
+  d->lengthResultLabel->clear();
+}
+
+//----------------------------------------------------------------------------
+void qSlicerStenosisMeasurement3DModuleWidget::SetOrCreateOutputTable()
```

Comment:
Your module should be usable without GUI. Therefore, essential features like creating columns in the output table must be implemented in the module logic class. You can use `vtkMRMLTr("SomeContext", "SomeMessage")` to get translations in VTK classes.

Reply:
Moved to logic.